reference/behavior_test/establishing_decorator_pattern.rst

```diff
@@ -0,0 +1,369 @@
+        }, sprintf('Something "%s" not found.', $foo));
+    }
+
+The method *$this->find()* is not in the decorator nor the abstract class ElementDecorator. The magical method
+*__call()* will be called to find in the Element if the method exist, and call it with arguments :
+
+.. code-block:: php
+
+    // abstract class Pim\\Behat\\Decorator\\ElementDecorator
+    /**
+     * @param string $name
+     * @param array  $arguments
+     *
+     * @return mixed
+     */
+    public function __call($name, array $arguments)
+    {
+        return call_user_func_array([$this->element, $name], $arguments);
+    }
+
+That's how we can call in decorators methods from the Element. And obviously, that's why all decorators must extend
+`Pim\\Behat\\Decorator\\ElementDecorator
+<https://github.com/akeneo/pim-community-dev/blob/master/features/Pim/Behat/Decorator/ElementDecorator.php>`_.
+
+.. note::
+
+    You can see that in the ElementDecorator a *decorate()* method exists. It is possible to use it to decorate an
+    Element in your decorator. In this example, it could be possible in the PermissionDecorator to create a
+    GroupPermissionDecorator to regroup all behavior about groups menu (the left menu in the previous picture) and a
+    ResourcePermissionDecorator to regroup all behavior about resources (the right panel in the previous picture).
+
+Now, what happen when *grantGroup()* is called from our Role Edit page.
+
+.. code-block:: php
+
+    /**
+     * @param string $group
+     */
+    public function grantGroup($group) // $group = 'Products'
+    {
+        $iconElement = $this->findGroupIcon($group);
+
+        if ($iconElement->hasClass('icon-remove') || $iconElement->hasClass('icon-circle')) {
+            $iconElement->click();
+        }
+    }
+
+    /**
+     * @param $group
+     *
+     * @return NodeElement
+     */
+    protected function findGroupIcon($group) // $group = 'Products'
+    {
+        $groupElement = $this->findGroup($group);
+
+        return $this->spin(function () use ($groupElement) {
+            return $groupElement->find('css', '.acl-group-permission-toggle');
+        }, sprintf('Group icon "%s" not found', $group));
+    }
+
+    /**
+     * @param string $group
+     *
+     * @return NodeElement
+     */
+    public function findGroup($group) // $group = 'Products'
+    {
+        return $this->spin(function () use ($group) {
+            return $this->find('css', sprintf('.tab-groups .tab:contains("%s")', $group));
+        }, sprintf('Group "%s" not found.', $group));
+    }
+
+.. note::
+
+    To know more about spin please refer to the `spin chapter <spin.html>`_.
+
+You may notice that all logic of the element research is located in decorators and are spinned.
+
+If we had several identical permission panels in different pages, each page has to use the same decorator. That's why,
+decorators have to be generic. In case an element is different between pages, use a decorator for those elements and
```

Comment:
Not necessarily @rybus , "In case an element is different" works as well :)
But I would say "In case an element differs from one page to another"